src/execution_plan/ops/op_delete.c

```diff
@@ -18,50 +19,73 @@ static OpResult DeleteReset(OpBase *opBase);
 static void DeleteFree(OpBase *opBase);
 
 void _DeleteEntities(OpDelete *op) {
-	Graph  *g                     =  op->gc->g;
-	uint   node_deleted           =  0;
 	uint   edge_deleted           =  0;
 	uint   implicit_edge_deleted  =  0;
 	uint   node_count             =  array_len(op->deleted_nodes);
 	uint   edge_count             =  array_len(op->deleted_edges);
 
 	// nothing to delete, quickly return
-	if((node_count + edge_count) == 0) goto cleanup;
+	if((node_count + edge_count) == 0) return;
 
-	// lock everything
-	QueryCtx_LockForCommit();
+	//--------------------------------------------------------------------------
+	// removing duplicates
+	//--------------------------------------------------------------------------
 
-	if(GraphContext_HasIndices(op->gc)) {
-		for(int i = 0; i < node_count; i++) {
-			Node *n = op->deleted_nodes + i;
-			GraphContext_DeleteNodeFromIndices(op->gc, n);
-		}
+	// remove node duplicates
+	Node *nodes = op->deleted_nodes;
+	Node *distinct_nodes = array_new(Node, 1);
 
-		for(int i = 0; i < edge_count; i++) {
-			Edge *e = op->deleted_edges + i;
-			GraphContext_DeleteEdgeFromIndices(op->gc, e);
-		}
+#define is_entity_lt(a, b) (ENTITY_GET_ID((a)) < ENTITY_GET_ID((b)))
+	QSORT(Node, nodes, node_count, is_entity_lt);
+
+	for(uint i = 0; i < node_count; i++) {
+		while(i < node_count - 1 && ENTITY_GET_ID(nodes + i) == ENTITY_GET_ID(nodes + i + 1)) i++;
+
+		if(DataBlock_ItemIsDeleted((nodes + i)->entity)) continue;
+		array_append(distinct_nodes, *(nodes + i));
+	}
+
+	node_count = array_len(distinct_nodes);
+
```

Comment:
I think it would be nice to collect all implicit edges of unique nodes here add them to `edges` and then sort and remove duplicates, basically reducing the logic from `GraphContext_DeleteNode`